pkg/avx/collector.go

```diff
@@ -77,111 +102,126 @@ var (
 	log = logger.NewLogger("avx")
 )
 
-func kernelVersionCode(major, minor, patch uint8) uint32 {
-	return uint32(major)<<16 + uint32(minor)<<8 + uint32(patch)
+type collector struct {
+	root string
+	ebpf *bpf.Collection
+	fds  []int
 }
 
-// getElfKernelVersion returns major, minor and patch parts of kernel version
-// compiled into eBPF ELF file.
-func getElfKernelVersion(path string) (uint8, uint8, uint8, error) {
-	elfFile, err := elf.Open(path)
+func enablePerfTracepoint(prog *bpf.Program, tracepoint string) (int, error) {
+
+	id, err := ioutil.ReadFile(filepath.Join(kernelTracepointPath, tracepoint, "id"))
+	if err != nil {
+		return -1, errors.Wrap(err, "unable to read tracepoint ID")
+	}
+
+	tid, err := strconv.Atoi(strings.TrimSpace(string(id)))
 	if err != nil {
-		return 0, 0, 0, errors.Wrapf(err, "unable to open ELF file %s", path)
+		return -1, errors.New("unable to convert tracepoint ID")
 	}
-	defer elfFile.Close()
 
-	sec := elfFile.Section("version")
-	if sec == nil {
-		return 0, 0, 0, errors.New("unable to find 'version' section")
+	attr := unix.PerfEventAttr{
+		Type:        unix.PERF_TYPE_TRACEPOINT,
+		Config:      uint64(tid), // tracepoint id
+		Sample_type: unix.PERF_SAMPLE_RAW,
+		Sample:      1,
+		Wakeup:      1,
 	}
-	data, err := sec.Data()
+
+	pfd, err := unix.PerfEventOpen(&attr, -1, 0, -1, unix.PERF_FLAG_FD_CLOEXEC)
 	if err != nil {
-		return 0, 0, 0, errors.Wrap(err, "unable to get version data")
+		return -1, errors.Wrap(err, "unable to open perf events")
 	}
 
-	// Least Significant Byte first
-	return data[2], data[1], data[0], nil
+	if _, _, errno := unix.Syscall(unix.SYS_IOCTL, uintptr(pfd), unix.PERF_EVENT_IOC_ENABLE, 0); errno != 0 {
```

Comment:
Hmm, is it defined what `errno` is if the call succeeds (_"On error, -1 is returned, and errno is set appropriately."_)? Shouldn't you check the syscall return value?

Reply:
`errno` is a `Syscall.Errno`: "the zero Errno is by convention a non-error" and I'm checking `!= 0`. Do you think the variable name (`errno`) is misleading?

Reply:
I think I'm still stuck to the syscall return value of thinking. :-) If you know that errno 0 means that the call is successful, then that's the way to go.